ibis/expr/scope.py

```diff
@@ -0,0 +1,182 @@
+        self._items[op] = value
+
+    def get(self, op: Node, timecontext: Optional[TimeContext] = None) -> Any:
+        """ Given a op and timecontext, get the result from scope
+
+        Parameters
+        ----------
+        scope : collections.Mapping
+            a dictionary mapping :class:`~ibis.expr.operations.Node`
+            subclass instances to concrete data, and the time context associate
+            with it (if any).
+        op: ibis.expr.operations.Node
+            key in scope.
+        timecontext: Optional[TimeContext]
+
+        Returns
+        -------
+        result: the cached result, an object whose types may differ in
+        different backends.
+        """
+        if op not in self.items():
+            return None
+
+        # for ops without timecontext
+        if timecontext is None:
+            return self.__getitem__(op).value
+        else:
+            # For op with timecontext, ther are some ops cannot use cached
+            # result with a different (larger) timecontext to get the
+            # correct result.
+            # For example, a groupby followed by count, if we use a larger or
+            # smaller dataset from cache, we will get an error in result.
+            # Such ops with global aggregation, ops whose result is
+            # depending on other rows in result Dataframe, cannot use cached
+            # result with different time context to optimize calculation.
+            # These are time context sensitive operations. Since these cases
+            # are rare in acutal use case, we just enable optimization for
+            # all nodes for now.
+            cached_timecontext = self.__getitem__(op).timecontext
+            if cached_timecontext:
+                relation = compare_timecontext(timecontext, cached_timecontext)
+                if relation == TimeContextRelation.SUBSET:
+                    return self.__getitem__(op).value
+            else:
+                return self.__getitem__(op).value
+        return None
+
+    def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
+        """merge items in other_scope into this scope
+
+        Parameters
+        ----------
+        other_scope: Scope
+            Scope to be merged with
+        overwrite: bool
+            if set to be True, force overwrite `value` if `op` already
+            exists.
+
+        Returns
+        -------
+        Scope
+            a new Scope instance with items in two scope merged.
+        """
+        result = Scope()
+
+        for op in self.items():
+            result.__setitem__(op, self.__getitem__(op))
+
+        for op in other_scope.items():
+            # if get_scope returns a not None value, then data is already
+            # cached in scope and it is at least a greater range than
+            # the current timecontext, so we drop the item. Otherwise
+            # add it into scope.
+            v = other_scope.__getitem__(op)
```

Comment:
I don't think you need to call __getitem__and __setitem.

`scope.__getitem__(op)` is just `scope[op]`
`scope.__setitem(op, value)` is just `scope[op] = value`

Reply:
+1, I fixed that